includes/Core/DI/DI_Container.php

```diff
@@ -0,0 +1,258 @@
+
+/**
+ * DI container class.
+ *
+ * @since n.e.x.t
+ */
+class DI_Container implements ContainerInterface, ArrayAccess {
+
+	/**
+	 * Definitions list.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $definitions = array();
+
+	/**
+	 * Returns true if the container can return an entry for the given identifier.
+	 * Returns false otherwise.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return bool TRUE if the container can return an entry, otherwise FALSE.
+	 */
+	public function offsetExists( $id ) {
+		return $this->has( $id );
+	}
+
+	/**
+	 * Finds an entry of the container by its identifier and returns it.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return mixed Entry.
+	 */
+	public function offsetGet( $id ) {
+		return $this->get( $id );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 */
+	public function offsetSet( $id, $entry ) {
+		if ( is_callable( $entry ) ) {
+			$this->set_service( $id, $entry );
+		} else {
+			$this->set_value( $id, $entry );
+		}
+	}
+
+	/**
+	 * Removes the entry.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to remove.
+	 */
+	public function offsetUnset( $id ) {
+		unset( $this->definitions[ $id ] );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param array  $entry Entry definition.
+	 * @return bool TRUE if the entry is added, otherwise FALSE.
+	 */
+	protected function set( $id, array $entry ) {
+		if ( ! empty( $this->definitions[ $id ]['is_protected'] ) ) {
+			return false;
+		}
+
+		$this->definitions[ $id ] = $entry;
+
+		return true;
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Entry name.
+	 * @param mixed  $value Entry value.
+	 * @return bool TRUE if the service is added, otherwise FALSE.
+	 */
+	public function set_value( $id, $value ) {
+		return $this->set(
+			$id,
+			array(
+				'entry' => $value,
+			)
+		);
+	}
+
+	/**
+	 * Sets the service definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string   $id Service name.
+	 * @param callable $create_func Service creator.
+	 * @return bool TRUE if the service is added, otherwise FALSE.
+	 */
+	public function set_service( $id, $create_func ) {
+		return $this->set(
+			$id,
+			array(
+				'is_service' => true,
+				'entry'      => $create_func,
+			)
+		);
+	}
+
+	/**
+	 * Sets the entry to be factory.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string   $id Identifier of the entry.
+	 * @param callable $factory_func Factory function.
+	 * @return bool TRUE if the factory is added, otherwise FALSE.
+	 */
+	public function set_factory( $id, $factory_func ) {
+		return $this->set(
+			$id,
+			array(
+				'is_service' => true,
```

Comment:
This shouldn't be set here, only `is_factory`.

Reply:
Updated.